geode-core/src/main/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessenger.java

```diff
@@ -179,6 +182,19 @@
    */
   private JGroupsReceiver jgroupsReceiver;
 
+  public static void setChannelReceiver(JChannel channel, Receiver r) {
+    try {
+      // Channel.setReceiver() won't set a new receiver unless the field is null but that
+      // causes NPEs in receiver threads because use of the the field isn't synchronized.
+      // Attempt to use reflection to avoid having the field be null. See GEODE-7220
+      Field receiver = Channel.class.getDeclaredField("receiver");
+      receiver.setAccessible(true);
+      receiver.set(channel, r);
```

Comment:
I don't understand how this change addresses the issue (of a thread getting an NPE). The field is neither `volatile` nor are accesses to it `synchronized`. Threads that have read a `null` value for this field may (in general) continue to see a `null` value for it, after this block completes.

Reply:
The state of the "receiver" field is established prior to starting the channel's reader threads.  What I'm trying to avoid is setting it to null again after that happens and also avoid a "warning" message, which will trigger an operator alert.
That we're hitting NPEs due to this field being modified with setReceiver(null) tells me that synchronization is happening somewhere BUT the jgroups code is using an anti-pattern of
reading the state of "receiver" in one method and then assuming that it hasn't changed state in a method that it invokes right after that read.

if (receiver != null) notifyReciever()
.
.
.
notifyReceiver() {
  receiver.receive(message); // NPE here
}

Reply:
... but I can see that there are no synchronizations between the two uses of "receiver", so we're just seeing the new state by chance.  That will be okay as long as it isn't null.  Both quorum-checker and jgroupsmessenger can respond to and consume "ping-pong" messages and startup retry logic can handle any other lost messages during the transition from one to the other.

Reply:
I can see how this avoids the warning message generated in `setReceiver(…)`. I don't see how this code avoids setting the field to `null` after starting the reader threads.

Reply:
JGroups threads will never see a null "receiver" with this code change because we establish a non-null "receiver" before starting JGroups threads.  The Thread.start() method is synchronized.  Thereafter we never set it to null again.

Reply:
ah is that due to the change in `GMSQuorumChecker.java`? (I hadn't seen that change, initially)